source/adaptive_load/adaptive_load_controller_impl.h

```diff
@@ -0,0 +1,80 @@
+#include "envoy/common/time.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_evaluator.h"
+#include "nighthawk/adaptive_load/session_spec_proto_helper.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+#include "nighthawk/common/nighthawk_service_client.h"
+
+namespace Nighthawk {
+
+class AdaptiveLoadControllerImpl : public AdaptiveLoadController {
+public:
+  /**
+   * Constructs an implementation of the adaptive load controller main loop that relies on logic in
+   * several helper objects. Through helpers, it performs Nighthawk Service benchmarks, obtains
+   * metrics from MetricsPlugins, scores the results, and consults a StepController plugin to
+   * determine the next load and detect convergence. All plugins are specified through the
+   * AdaptiveLoadSessionSpec proto.
+   *
```

Comment:
can you add a comment on the thread safety? Can you run concurrent sessions?

Reply:
Done